### PR TITLE
[Serve] add `RAY_SERVE_LOG_ENCODING` env to set the global logging behavior for Serve

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

We want a way to be able to use environment variables to setup Serve's structured json logging on the platform. Add `RAY_SERVE_LOG_ENCODING_ENV` env to set the global logging behavior for Serve.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/constants.py [hunk 1]
```diff
@@ -163,6 +163,8 @@
     "instead (see https://docs.ray.io/en/latest/serve/model_composition.html)."
 )
 
+# Environment variable name for to specify the encoding of the log messages
+RAY_SERVE_LOG_ENCODING_ENV = "RAY_SERVE_LOG_ENCODING_ENV"
```
Comment (edoakes) [inline python/ray/serve/_private/constants.py:167]: Drop the `_ENV` suffix; we don't use that for any of the other environment variable flags

File: python/ray/serve/tests/test_logging.py [hunk 2]
```diff
@@ -578,5 +585,52 @@ def format_and_verify_json_output(record, expected_record: dict):
...
+@pytest.mark.parametrize(
+    "log_encoding",
+    [
+        [None, "TEXT"],
+        ["TEXT", "TEXT"],
+        ["JSON", "JSON"],
+        ["FOOBAR", "TEXT"],
+    ],
+)
+def test_configure_component_logger_with_log_encoding_env(monkeypatch, log_encoding):
```
Comment (edoakes) [inline python/ray/serve/tests/test_logging.py:597]: please also test the override behavior if users specify the encoding in their `logging_config`. I assume that env var is lowest precedence

File: python/ray/serve/_private/logging_utils.py [hunk 3]
```diff
@@ -228,8 +231,10 @@ def record_factory(*args, **kwargs):
         logs_dir = get_serve_logs_dir()
     os.makedirs(logs_dir, exist_ok=True)
 
-    max_bytes = ray._private.worker._global_node.max_bytes
-    backup_count = ray._private.worker._global_node.backup_count
+    if not max_bytes:
+        max_bytes = ray._private.worker._global_node.max_bytes
+    if not backup_count:
```
Comment (edoakes) [inline python/ray/serve/_private/logging_utils.py:236]:
```suggestion
    if backup_count is None:
```
Have to be careful -- `backup_count=0` is valid

File: python/ray/serve/_private/logging_utils.py [hunk 4]
```diff
@@ -228,8 +231,10 @@ def record_factory(*args, **kwargs):
         logs_dir = get_serve_logs_dir()
     os.makedirs(logs_dir, exist_ok=True)
 
-    max_bytes = ray._private.worker._global_node.max_bytes
-    backup_count = ray._private.worker._global_node.backup_count
+    if not max_bytes:
```
Comment (edoakes) [inline python/ray/serve/_private/logging_utils.py:234]:
```suggestion
    if max_bytes is None:
```

File: python/ray/serve/tests/test_logging.py (hunk 2 above)
Comment (GeneDer) [reply python/ray/serve/tests/test_logging.py:597]: Makes sense, but this kinda requires us to default to an encoding that's neither text nor json in the `logging_config` so we know if user has explicitly set it. Let me take a stab

Comment (GeneDer) [reply python/ray/serve/tests/test_logging.py:605]: Right now just use `None` to represent the encoding not being set in the logging config. But open to adding new enum of `NOT_SET` or other ideas :)

Comment (GeneDer) [reply python/ray/serve/tests/test_logging.py:605]: Ouch, this actually won't work. The protobuf doesn't like `None`😅 Gonna just add a `NOT_SET` enum

File: python/ray/serve/_private/constants.py [hunk 5]
```diff
@@ -163,6 +163,8 @@
     "instead (see https://docs.ray.io/en/latest/serve/model_composition.html)."
 )
 
+# Environment variable name for to specify the encoding of the log messages
+RAY_SERVE_LOG_ENCODING = os.environ.get("RAY_SERVE_LOG_ENCODING", "TEXT")
 # Jsonify the log messages
 RAY_SERVE_ENABLE_JSON_LOGGING = os.environ.get("RAY_SERVE_ENABLE_JSON_LOGGING") == "1"
```
Comment (edoakes) [inline python/ray/serve/_private/constants.py:169]: note here that this one is deprecated and getting removed

File: python/ray/serve/schema.py [hunk 6]
```diff
@@ -130,10 +135,11 @@ class Config:
         extra = Extra.forbid
 
     encoding: Union[str, EncodingType] = Field(
-        default="TEXT",
+        default_factory=_ray_serve_log_encoding,
```
Comment (edoakes) [inline python/ray/serve/schema.py:138]:
```suggestion
        default_factory=lambda: RAY_SERVE_LOG_ENCODING,
```
A little easier to read IMO :)

File: python/ray/serve/schema.py [hunk 7]
```diff
@@ -130,10 +135,11 @@ class Config:
...
         description=(
-            "Encoding type for the serve logs. Default to 'TEXT'. 'JSON' is also "
-            "supported to format all serve logs into json structure."
+            "Encoding type for the serve logs. Default to 'TEXT'. The default can be "
```
Comment (edoakes) [inline python/ray/serve/schema.py:140]:
```suggestion
            "Encoding type for the serve logs. Defaults to 'TEXT'. The default can be "
```

File: python/ray/serve/schema.py [hunk 8]
```diff
@@ -130,10 +135,11 @@ class Config:
...
+            "overwritten using the `RAY_SERVE_LOG_ENCODING` environment variable. "
+            "'JSON' is also supported to format all serve logs into json structure."
```
Comment (edoakes) [inline python/ray/serve/schema.py:142]:
```suggestion
            "'JSON' is also supported for structured logging."
```

**Review by edoakes — COMMENTED** [review]
Just small comments

File: python/ray/serve/schema.py (hunk 6 above)
Comment (edoakes) [reply python/ray/serve/schema.py:138]: I think this can actually just be `RAY_SERVE_LOG_ENCODING`, no need for the factory

Comment (GeneDer) [reply python/ray/serve/schema.py:138]: Yea, I tried `default= RAY_SERVE_LOG_ENCODING` initially, but patching and reloading modules correctly in pytest is almost impossible 😅

Comment (edoakes) [reply python/ray/serve/schema.py]: Got it 👍

**Review by edoakes — APPROVED** [review]

**Comment by GeneDer** [comment]
@edoakes this is ready for merge

**Pull request merged**